### DISCUSSION TITLE
Stat extension recipes and cookbook

### DISCUSSION THREAD

**Discussion by EvaMaeRey** [ggplot2-extenders/ggplot-extension-club#27] [Show and tell]
I asked R and ggplot users to try 'easy geom recipes' tutorial more than a year ago.  

lightly revised in 2024: https://evamaerey.github.io/mytidytuesday/2022-01-03-easy-geom-recipes/easy_geom_recipes_flipbook_2024.html

![image](https://github.com/teunbrand/ggplot-extension-club/assets/10491910/7dca9aa8-9437-42d0-8259-fcc2f2205273)

A small group did the tutorial and completed a survey about their background and experience with the survey:  https://evamaerey.github.io/easy-geom-recipes/survey_results_summary.html

Here's the distribution for one of the survey questions:

![image](https://github.com/teunbrand/ggplot-extension-club/assets/10491910/d704ada8-92c7-460f-abe7-fcf5718c03e7)

They did great, but one participant commented in a focus group:

> 'it was that easy. And I felt empowered as a result of that…. But you know, like, my problem isn’t gonna be that easy.'.  

Yup, true true.  The introductory tutorial should gives a taste of success, but you might be stymied if you want something that doesn't quite parallel the three very introductory recipes.  So, collecting more Stat building experiences in this cookbook that could possibly help out (mirrors the step-by-step approach of the original tutorial).  

Thanks to everyone here that has contributed so greatly to filling in my know-how gaps - which have been many and still are many.    

I welcome thoughts/concerns/feedback/reservations on the project. 

https://evamaerey.github.io/ggplot2-extension-cookbook/

![image](https://github.com/teunbrand/ggplot-extension-club/assets/10491910/d0ee7ed9-fe53-4626-9da0-52bdd59e2541)

**Comment by EvaMaeRey** [comment]
I'm probably going to use this language in the recipe/cookbook for writing classical layer()-base user-facing stat_*() function.  I think getting the 'boilier plate'/'scaffolding' for the user facing function can be tricky.  Personally, the recommendation below only recently occurred to me, and before I'd always have to depart to a vignette or previous tutorial that I'd written (so like google search or file search) and it felt messy.  This protocol avoids that.  Thoughts?  @yjunechoe

> Protip: Use `stat_identity()`'s definition as scaffolding to write your own layer. i.e. 1. type `stat_identity` in your console 2. copy-paste function definition. 3. Switch out StatIdentity with your Stat and Geom if not 'point'. 4 Finally, list2() will error without export from rlang, so update to rlang::list2().

**Comment by teunbrand** [comment]
I've opened a [PR](https://github.com/tidyverse/ggplot2/pull/6103) for adding a 'manual' stat, where essentially the `compute_group()` method is available to the user.
Probably a lot of your recipes work there as well, without having to actually build a stat :)

**Reply by EvaMaeRey** [reply]
hmm!

**Reply by EvaMaeRey** [reply]
Sounds like this does lots that #26, #31 is aiming at.    Are you thinking about panel-wise, layer-wise compute as well?

**Reply by teunbrand** [reply]
It is an interesting notion, but I don't think I know of any examples of where computing stats by panel or layer is convenient.

**Reply by EvaMaeRey** [reply]
I usually do so when computation happens across discrete variables... 

Maybe a linear y ~ x + z (where z is categorical - and would break data into groups)
https://evamaerey.github.io/ggplot2-extension-cookbook/#geom_ols_linear_parallel-nkw-interdependence

I think the waterfall plot and circlepacking can't happen without moving up from group-wise computation, because the variables defining event on x axis for waterfall and circle id are usually factor or character.

https://evamaerey.github.io/ggplot2-extension-cookbook/#geom_waterfall-111-compute_panel-geomrect
https://evamaerey.github.io/ggplot2-extension-cookbook/#geom_circlepack-11n-compute_panel-geompolygon

StatPairs - x might be discrete and pair id will probably be discrete.
https://github.com/EvaMaeRey/ggplot2-extension-cookbook/discussions/2#discussioncomment-10559014

StatTextcircle - label var is discrete, but placement depends on knowing about other vars.  But I don't remember if label actually triggers group-wise computation.
https://github.com/nrennie/ggtextcircle/blob/main/R/StatTextcircle.R

**Reply by teunbrand** [reply]
Thanks for all these examples! I think a common theme of these is that they have 1:1 input:output correspondance, like the semantics of `dplyr::mutate()` Thereby, these computations preserve any existing aesthetics, like `group` or `PANEL` that might otherwise get lost. In such cases, you'd probably want to manually set groups instead of relying on the default ones to get the computation correct.

For 1:many this is probably still fine, but for many:1 or many:many, like `dplyr::summarise()` or `dplyr::reframe()` this probably won't work out as relevant columns might be dropped by accident, and there are no mechanisms in place to preserve such columns (whereas this mechanism exists for group-level computation).

I wouldn't really trust the average user to handle these nuances without running into frustrations, so I think the safer thing is to just allow group level computation.

**Reply by EvaMaeRey** [reply]
I'm unclear if you have suggestions on modification for the examples that are provided or not.  If yes, would you mind providing some detail one what you think should be changed for one of the examples?   Somehow I have never heard of `reframe` before this week!  

I think that a #31 `StatH <- ggproto("StatH", Stat, compute_group = makehull);  geom_polygon(stat =StatH)` approach might be more open ended (like later adding default aes) and internals-illuminating.  statexpress::stat_group() in #26 was *meant* to work like stat_manual() but expose more ways to define the Stat, like default_aes and required_aes etc, and statexpress::stat_panel exposed compute_panel. I don't know if it can actually work.   

I guess I'll need to add a call out on stat_manual to the recipes when next ggplot2 release is out!

**Reply by teunbrand** [reply]
No I don't have suggestions for the examples; they're perfectly fine as-is. I mean your audience at that point must be mostly people that think about writing their own stats, which is definitely more advanced that the average ggplot2 user.

I can see the similarities with the statexpress functions, except `stat_manual()` is more a dumbed down version 😅

**Reply by yjunechoe** [reply]
Learning a lot from the discussion here! Just to chime in with a personal experience:

> I mean your audience at that point must be mostly people that think about writing their own stats, which is definitely more advanced that the average ggplot2 user.

It actually took me a very long time to develop an intuition for what it means for a stat to *not* compute by group - I'd always just assumed things happen at group-level because it's an abstraction that works in most contexts, and couldn't imagine an alternative.

The pivotal moment was later in my ggplot internals discovery phase when I realized that [StatUnique computed by panel](https://yjunechoe.github.io/posts/2022-07-06-ggplot2-delayed-aes-2/#other-compute_-extensions) - it's such a trivial case but took me a while to wrap my head around that (probably makes for a great teaching example, though!).

So maybe a litmus test for whether someone's ready to use something like a Panel-level stat_manual() would be whether they can explain the motivation and implementation of `stat = "unique"` 😆

**Reply by teunbrand** [reply]
Huh interesting, I hadn't really thought about StatUnique computing by panel. I think it might also be possible to compute the entire layer at once, as the the `PANEL` variable already makes observations in different panels unique. However, using `compute_layer()` to actually do the stat computation doesn't really happen in vanilla ggplot2.

**Reply by yjunechoe** [reply]
Yeah I think the implementation could be simplified but I like the cleverness of the current design where you directly take the logic of StatUnique (do a `unique(data)`) and the question is at which compute stage should that happen. I find it kind of charming!

**Reply by EvaMaeRey** [reply]
I thought about a different approach which might be more aligned to full Stat extension thinking and composition with new Stats and existing geoms:  

> `geom_polygon(stat = qstat(fun = make_hull))`  # qstat (quick stat) function wraps ggproto("StatTemp", Stat, compute_group = fun)


I think you could get back to a stat_manual(fun = make_hull) with the `qstat` working internally but also having it as an exported function?

Feeling much more attractive than what's currently statexpress at least (more concise and manageable?) There might be some holes in the thinking/execution that I'm not seeing though... 🤷‍♀️  

https://evamaerey.github.io/mytidytuesday/2024-09-19-temp-stat/temp-stat.html

``` r
library(tidyverse)

# Given a cool function ...
make_hull <- function(data, scales) {
  
  hull_ids <- chull(data %>% select(x,y))

  data |>
    dplyr::slice(hull_ids)
    
}

mtcars %>% 
  select(x = disp, y = mpg) %>% 
  make_hull()
#>                         x    y
#> Pontiac Firebird    400.0 19.2
#> Chrysler Imperial   440.0 14.7
#> Cadillac Fleetwood  472.0 10.4
#> Lincoln Continental 460.0 10.4
#> Merc 280C           167.6 17.8
#> Volvo 142E          121.0 21.4
#> Datsun 710          108.0 22.8
#> Fiat X1-9            79.0 27.3
#> Toyota Corolla       71.1 33.9
```

``` r


stat_manual <- statexpress::stat_group # 

# 1. stat_manual promise
library(ggplot2)
ggplot(mtcars, aes(disp, mpg, colour = factor(cyl))) +
  geom_point() + 
  stat_manual(fun = make_hull, geom = "polygon", alpha = .2) + 
  stat_manual(fun = make_hull, size = 3)
```

![](https://i.imgur.com/OE2EM6l.png)

``` r

# 2. Express in-script dependency free approach
library(ggplot2)
StatHull <- ggproto("StatHull", Stat, compute_group = make_hull)
ggplot(mtcars, aes(disp, mpg, colour = factor(cyl))) +
  geom_point() + 
  geom_polygon(stat = StatHull, alpha = .2) + 
  geom_point(stat = StatHull, size = 3)
```

![](https://i.imgur.com/tjDEcaE.png)

``` r



# statexpress/stat_manual alternative... qstat
qstat <- function(fun, ...){ggproto("StatTemp", Stat, compute_group = fun, ...)}
qstat_p <- function(fun, ...){ggproto("StatTemp", Stat, compute_panel = fun, ...)}

# 3. third way: qstat
library(ggplot2) # or library(statexpress)
ggplot(mtcars, aes(disp, mpg, colour = factor(cyl))) +
  geom_point() + 
  geom_polygon(stat = qstat(make_hull), alpha = .2) + 
  geom_point(stat = qstat(make_hull), size = 3)
```

![](https://i.imgur.com/3dV5Akg.png)

``` r

# 3.b can still define a stat, but subclass is 'StatTemp' internally
library(ggplot2)
QSH <- qstat(make_hull)
ggplot(mtcars, aes(disp, mpg, colour = factor(cyl))) +
  geom_point() + 
  geom_polygon(stat = QSH, alpha = .2) + 
  geom_point(stat = QSH, size = 3)
```

![](https://i.imgur.com/G2sTUm2.png)

``` r


# Niceties of new approach
# Free to pack your computation w/ bonus columns ...
make_hull_w_means <- function(data, scales){
 
  data %>% make_hull() %>% 
    mutate(hull_mean_x = mean(x)) %>% 
    mutate(hull_mean_y = mean(y)) %>% 
    mutate(hull_median_x = median(x)) %>% 
    mutate(hull_median_y = median(y))
  
}


# ... for use in your stat... 
QStatHull <- qstat(fun = make_hull_w_means, 
                   default_aes = aes(xend = after_stat(hull_mean_x),
                                     yend = after_stat(hull_mean_y)))

last_plot() + 
  geom_segment(stat = QStatHull) + 
  labs(title = "the means based my c-hull...")
```

![](https://i.imgur.com/a5lB7fd.png)

``` r

last_plot() + 
  geom_segment(stat = QStatHull, xend = 0, yend = 0, alpha = .2) + 
  labs(caption = "let's anchor these to the origin!")
```

![](https://i.imgur.com/pMR1oMk.png)

``` r


# using group means
group_center <- function(data, scales, fun = mean){
  
 data |> 
    summarise(x = fun(x),
              y = fun(y))
  
}

last_plot() + 
  geom_point(stat = qstat(group_center), 
             size = 7,
             shape = "diamond") + 
  labs(subtitle = "... is not the group means")
```

![](https://i.imgur.com/1R7teHF.png)

``` r
  

last_plot() + 
  geom_label(stat = qstat(group_center), 
             label = "global\nmeans",
             color = "black",
             shape = "diamond")
#> Warning in geom_label(stat = qstat(group_center), label = "global\nmeans", :
#> Ignoring unknown parameters: `shape`
```

![](https://i.imgur.com/d2OfFox.png)

``` r

last_plot() + 
  geom_text(stat = qstat(group_center), 
            label = "global\nmedians",
            color = "black",
            fun = median)
```

![](https://i.imgur.com/YEVkZAD.png)

``` r


# when you don't have a stat argument for geom_*() as with geom_vline()
# cross between geom_point and stat_identity
layer_express <- function (mapping = NULL, data = NULL, stat = "identity", geom = "point", position = "identity", 
    ..., na.rm = FALSE, show.legend = NA, inherit.aes = TRUE) {
    layer(data = data, mapping = mapping, stat = stat, geom = geom, 
        position = position, show.legend = show.legend, inherit.aes = inherit.aes, 
        params = rlang::list2(na.rm = na.rm, ...))
} 

xmean <- function(data, scales){
  
 data |> 
    summarise(xintercept = mean(x))
  
}


last_plot() +
  layer_express(stat = qstat(xmean, dropped_aes = c("x", "y")), 
                geom = GeomVline)
```

![](https://i.imgur.com/fxWyB98.png)

``` r

# And back again!  should be same behavior as stat_manual()
stat_manual_group <- function(geom = "point", group_fun, ...){
  
  layer_express(geom = geom, stat = qstat(fun = group_fun), ...)
  
}




ggplot(mtcars, aes(disp, mpg, colour = factor(cyl))) +
  geom_point() + 
  stat_manual_group(group_fun = make_hull, geom = "polygon", alpha = .2) + 
  stat_manual_group(group_fun = make_hull, size = 3)
```

![](https://i.imgur.com/B5wpoPK.png)

<sup>Created on 2024-09-19 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>
